File: ios/Shared/TwoBeolsikLayout.swift
enum TwoBeolsikLayout {
    static let characterRows: [[String]] = [
        ["ㅂ", "ㅈ", "ㄷ", "ㄱ", "ㅅ", "ㅛ", "ㅕ", "ㅑ", "ㅐ", "ㅔ"],
        ["ㅁ", "ㄴ", "ㅇ", "ㄹ", "ㅎ", "ㅗ", "ㅓ", "ㅏ", "ㅣ"]
    ]

    static let bottomRow: [String] = ["ㅋ", "ㅌ", "ㅊ", "ㅍ", "ㅠ", "ㅜ", "ㅡ"]

    private static let shiftedKeys: [String: String] = [
        "ㅂ": "ㅃ",
        "ㅈ": "ㅉ",
        "ㄷ": "ㄸ",
        "ㄱ": "ㄲ",
        "ㅅ": "ㅆ",
        "ㅐ": "ㅒ",
        "ㅔ": "ㅖ"
    ]

    static func label(for baseLabel: String, shifted: Bool) -> String {
        guard shifted else { return baseLabel }
        return shiftedKeys[baseLabel] ?? baseLabel
    }

    static func hasShiftVariant(_ baseLabel: String) -> Bool {
        shiftedKeys[baseLabel] != nil
    }
}

File: android/app/src/main/java/com/ghtnql/kkkeyboard/TwoBeolsikLayout.kt
package com.ghtnql.kkkeyboard

object TwoBeolsikLayout {
    val auxiliaryNumberRow: List<String> = listOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "0")

    val characterRows: List<List<String>> = listOf(
        listOf("ㅂ", "ㅈ", "ㄷ", "ㄱ", "ㅅ", "ㅛ", "ㅕ", "ㅑ", "ㅐ", "ㅔ"),
        listOf("ㅁ", "ㄴ", "ㅇ", "ㄹ", "ㅎ", "ㅗ", "ㅓ", "ㅏ", "ㅣ"),
    )

    val bottomRow: List<String> = listOf("ㅋ", "ㅌ", "ㅊ", "ㅍ", "ㅠ", "ㅜ", "ㅡ")

    private val shiftedKeys = mapOf(
        "ㅂ" to "ㅃ",
        "ㅈ" to "ㅉ",
        "ㄷ" to "ㄸ",
        "ㄱ" to "ㄲ",
        "ㅅ" to "ㅆ",
        "ㅐ" to "ㅒ",
        "ㅔ" to "ㅖ",
    )

    fun labelFor(baseLabel: String, shifted: Boolean): String =
        if (shifted) shiftedKeys[baseLabel] ?: baseLabel else baseLabel

    fun hasShiftVariant(baseLabel: String): Boolean = shiftedKeys.containsKey(baseLabel)
}
